Colorize dendrogram

stefanhahmann opened the pull request:
Adds colorizing the dendrogram with the same colors that are used for the tag sets

Resolves #49

maarzt commented on src/test/java/org/mastodon/mamut/clustering/ui/DendrogramPanelDemo.java:
```diff
@@ -0,0 +1,49 @@
+import java.awt.Color;
+import java.util.Collections;
+
+public class DendrogramPanelDemo
```
A javadoc comment that describes what the class does would be awesome

stefanhahmann replied:
Done here: https://github.com/mastodon-sc/mastodon-deep-lineage/commit/bc81b778832ee031aad1f465ae773754cef10acb

maarzt commented on src/test/java/org/mastodon/mamut/clustering/ClusterApporientedDemo.java:
```diff
@@ -8,10 +8,12 @@
+import java.util.Collections;
 import java.util.Random;
 
 public class ClusterApporientedDemo
```
A javadoc comment that describes what the class does would be awesome

stefanhahmann replied:
Done here: https://github.com/mastodon-sc/mastodon-deep-lineage/commit/4ee01dbb286adfed644d23b152dc9b523d53b7d4

maarzt reviewed (commented):
Nice functionality. I tested the "cluster lineage tree" plugin and it works as expected. 
The test coverage is impressive. Which gives me confidence that the code will run reliably.

The code quality is goof. But I see one mayor point for improvement. This PR adds new methods to the "Classification" class. The class became a bit unorganized. I think restructuring it would be valuable.

Regarding javadoc & comments: There are sometimes concepts or surprises in the code. That are nowhere described, for example:
* The class ID in the Classification class, what does it refer to.
* We no the the "WeightedLinkageStrategy" class is named wrongly. And therefor we decided extend the class. This should be described in the commet.
* That a class is a copy of a class from another repository.
* There are  tests that only test a side effect "CustomizedClusterComponentTest"

Please add, one sentence of javadoc to each class in the repository. One sentence that  describes the purpose of class.

There is a small problem the color of the 4th class is always black. This is due to the use of the glasby lookup table and the fact that the first 5 colors in there are "outliers colors". I would suggest to exclude the first five colors.

stefanhahmann commented on src/main/java/org/mastodon/mamut/clustering/ui/DendrogramPanel.java:
```diff
@@ -0,0 +1,345 @@
+		return g.getFontMetrics().getHeight() + 2 * SCALE_PADDING + SCALE_TICK_LABEL_PADDING;
+	}
+
+	class ScaleBar
```
Good advice. Addressed here: https://github.com/mastodon-sc/mastodon-deep-lineage/commit/b8e2fdf6e1298ea85033445c0fe8d8066b17a358

stefanhahmann commented on src/test/java/org/mastodon/mamut/clustering/ui/DendrogramPanelTest.java:
```diff
@@ -0,0 +1,125 @@
+						3
+				);
+
+		graphics = Mockito.mock( Graphics2D.class );
```
Done here: https://github.com/mastodon-sc/mastodon-deep-lineage/commit/31d243b9373f713c7d58c9984b132f734d7a31df
In fact new Graphics2D would require a quite lengthy bit of code to implement this the abstract class Graphics2D

stefanhahmann commented on src/test/java/org/mastodon/mamut/clustering/ui/CustomizedClusterComponentTest.java:
```diff
@@ -0,0 +1,45 @@
+		graphics.setColor( defaultColor );
+
+		customizedClusterComponent.paint( ( Graphics2D ) graphics, 0, 0, 1d, 1d, false );
+		assertEquals( defaultColor, graphics.getColor() );
```
Done here: https://github.com/mastodon-sc/mastodon-deep-lineage/commit/b7e62033928b1b22c058395341eb6f00ac70523b

stefanhahmann commented on src/main/java/org/mastodon/mamut/clustering/ui/CustomizedClusterComponent.java:
```diff
@@ -0,0 +1,78 @@
+ *
+ * @author Stefan Hahmann
+ */
+public class CustomizedClusterComponent< T > extends ClusterComponent
```
Thank you for this advice. I addressed it here: https://github.com/mastodon-sc/mastodon-deep-lineage/commit/7119f70cf04493fc9217adc03e77331aeb88761e

stefanhahmann commented on src/main/java/org/mastodon/mamut/clustering/ui/CustomizedClusterComponent.java:
```diff
@@ -0,0 +1,78 @@
+/**
+ * Class that extends {@link ClusterComponent} to customize the color of the dendrogram part represented by it.
+ *
+ * @author Stefan Hahmann
```
Addressed here: https://github.com/mastodon-sc/mastodon-deep-lineage/commit/27c9f84b810a2059ad3f907772e387933c22b754

stefanhahmann commented on src/main/java/org/mastodon/mamut/clustering/util/Classification.java:
```diff
@@ -1,34 +1,68 @@
+import java.util.stream.Collectors;
 
+/**
+ * A class that encapsulates the result of a clustering algorithm.<p>
+ *     It contains:
+ *     <ul>
+ *         <li>an ordered list of classified objects</li>
+ *         <li>the result of the clustering algorithm</li>
+ *         <li>a mapping from leaf names to objects</li>
+ *         <li>the cutoff value of classification, i.e. where the dendrogram is cut</li>
+ *         <li>the colors of the clusters and the class ids</li>
+ *     </ul>
+ * @author Stefan Hahmann
```
I restructured the class `Classification` in accordance to your suggestion.

That allowed:

- getting rid of the obscure "leaf name" variable
- getting rid of any cluster ids

I renamed `algorithmResult` to `rootCluster`, since this is indeed more appropriate.

Cf.: https://github.com/mastodon-sc/mastodon-deep-lineage/commit/3dead27ece04456c751d8526428dec20b13ec1db

stefanhahmann commented on src/main/java/org/mastodon/mamut/clustering/ui/DendrogramPanel.java:
```diff
@@ -0,0 +1,345 @@
+ * Class for painting dendrograms derived from a {@link Classification} object.<p>
+ * In addition to {@link com.apporiented.algorithm.clustering.visualization.DendrogramPanel}, this class adds the possibility to paint an extra line on top of dendrogram (e.g. the cut-off line) and colorizing the clusters below this cut-off line.
+ *
+ * @author Stefan Hahmann
```
I would not call it a copy anymore, since the changes compared to the original class a quite substantial. I called it a "re-implementation now". I added the differences to the original class more carefully now. Cf.: https://github.com/mastodon-sc/mastodon-deep-lineage/commit/ed1c892afe79f9000399950b993c7639e23b4389

stefanhahmann commented:
> Nice functionality. I tested the "cluster lineage tree" plugin and it works as expected. The test coverage is impressive. Which gives me confidence that the code will run reliably.
> 
> The code quality is goof. But I see one mayor point for improvement. This PR adds new methods to the "Classification" class. The class became a bit unorganized. I think restructuring it would be valuable.
> 
> Regarding javadoc & comments: There are sometimes concepts or surprises in the code. That are nowhere described, for example:
> 
>     * The class ID in the Classification class, what does it refer to.
> 
>     * We no the the "WeightedLinkageStrategy" class is named wrongly. And therefor we decided extend the class. This should be described in the commet.
> 
>     * That a class is a copy of a class from another repository.
> 
>     * There are  tests that only test a side effect "CustomizedClusterComponentTest"
> 
> 
> Please add, one sentence of javadoc to each class in the repository. One sentence that describes the purpose of class.
> 
> There is a small problem the color of the 4th class is always black. This is due to the use of the glasby lookup table and the fact that the first 5 colors in there are "outliers colors". I would suggest to exclude the first five colors.

Thanks for the review.

I did the following steps:

- restructured the Classification class so that the IDs are no longer needed
- Added a comment explaining that WeightedLinkageStrategy was overridden, because it is wrong
- Added comments to the copied (or rather re-implemented classes)
- Explained that CustomizedClusterComponentTest has only limited testing value
- Added a javadoc to each new class
- Skipped the first five colors of the Glasbey color pallette.
